File: sdk/rust/src/catalog/event.rs
use std::sync::Arc;

use crate::{Code, Error, Fields, Result};

use super::model::{Kind, MAX_FIELDS, Path, array_field_index, canonical_u64, canonical_usize, validate_patch, validate_value};

#[derive(Clone, Copy)]
pub(super) enum EventKind {
    Replace,
    Patch,
    Delete,
}

pub(super) struct CatalogEvent {
    pub kind: EventKind,
    pub path: Path,
    pub revision: u64,
    pub base_revision: u64,
    pub value_kind: Option<Kind>,
    pub encoded_bytes: usize,
    pub fields: Arc<Fields>,
}

struct EventDecoder<'a> {
    payload: &'a [u8],
    offset: usize,
}

/// 解码并完整校验一条 Catalog Pub/Sub `payload`。
///
/// `expected` 用于拒绝错误频道或成员，`maximum` 同时约束通知和完整值大小。
/// 成功只返回通过协议版本、操作形状、revision、字段顺序和容量校验的拥有型事件。
pub(super) fn decode_event(payload: &[u8], expected: &Path, maximum: usize) -> Result<CatalogEvent> {
    // 在读取任意 MessagePack 容器前应用保守线长上限，避免异常负载消耗过多解析工作。
    if payload.len() > maximum_event_payload(maximum) {
        return Err(Error::field(Code::Capacity, "notification"));
    }
    let mut decoder = EventDecoder { payload, offset: 0 };
    let count = decoder.array_length().ok_or_else(|| Error::field(Code::Corrupt, "notification"))?;
    if count < 4 {
        return Err(Error::field(Code::Corrupt, "notification"));
    }
    if decoder.text("protocol")? != "v1" {
        return Err(Error::field(Code::Protocol, "protocol"));
    }
    let operation = decoder.text("operation")?;
    if !expected.matches_member(decoder.text("path")?) {
        return Err(Error::field(Code::Target, "path"));
    }

    // 公共头部确认后按固定操作 ABI 解码；每个分支都必须精确匹配数组元素数。
    let event = match operation {
        "replace" if count == 7 => {
            let revision = decode_event_revision(&mut decoder, "@revision")?;
            let kind = Kind::parse(decoder.text("@kind")?).ok_or_else(|| Error::field(Code::Corrupt, "@kind"))?;
            let encoded_bytes = decode_event_usize(&mut decoder, maximum, "@encoded_bytes")?;
            let fields = Arc::new(decode_event_fields(&mut decoder, kind == Kind::Array)?);
            if validate_value(kind, fields.as_ref(), maximum)? != encoded_bytes {
                return Err(Error::field(Code::Corrupt, "@encoded_bytes").with_revision(revision));
            }
            CatalogEvent {
                kind: EventKind::Replace,
                path: expected.clone(),
                revision,
                base_revision: 0,
                value_kind: Some(kind),
                encoded_bytes,
                fields,
            }
        }
        "patch" if count == 8 => {
            let base_revision = decode_event_revision(&mut decoder, "@base_revision")?;
            let revision = decode_event_revision(&mut decoder, "@revision")?;
            if revision <= base_revision {
                return Err(Error::field(Code::Corrupt, "@revision"));
            }
            let kind = Kind::parse(decoder.text("@kind")?)
                .filter(|kind| *kind != Kind::Value)
                .ok_or_else(|| Error::field(Code::Corrupt, "@kind"))?;
            let encoded_bytes = decode_event_usize(&mut decoder, maximum, "@encoded_bytes")?;
            let fields = Arc::new(decode_event_fields(&mut decoder, false)?);
            validate_patch(fields.as_ref(), maximum)?;
            CatalogEvent {
                kind: EventKind::Patch,
                path: expected.clone(),
                revision,
                base_revision,
                value_kind: Some(kind),
                encoded_bytes,
                fields,
            }
        }
        "delete" if count == 4 => CatalogEvent {
            kind: EventKind::Delete,
            path: expected.clone(),
            revision: decode_event_revision(&mut decoder, "@revision")?,
            base_revision: 0,
            value_kind: None,
            encoded_bytes: 0,
            fields: Arc::new(Fields::new()),
        },
        "replace" | "patch" | "delete" => {
            return Err(Error::field(Code::Corrupt, "notification"));
        }
        _ => return Err(Error::field(Code::Protocol, "operation")),
    };
    if !decoder.done() {
        return Err(Error::field(Code::Corrupt, "notification"));
    }
    Ok(event)
}

/// 计算 `maximum` 值上限对应的最坏通知线长。
///
/// 返回值额外容纳字段名称、MessagePack 长度头与协议元数据，并使用饱和算术防溢出。
fn maximum_event_payload(maximum: usize) -> usize {
    maximum.saturating_add(maximum.min(MAX_FIELDS).saturating_mul(10)).saturating_add(1024)
}

/// 从 `decoder` 读取 `field` 对应的规范正 revision。
fn decode_event_revision(decoder: &mut EventDecoder<'_>, field: &str) -> Result<u64> {
    let text = decoder.text(field)?;
    canonical_u64(text, false, field)
}

/// 从 `decoder` 读取一个不超过 `maximum` 的规范 `usize`，并以 `field` 定位错误。
fn decode_event_usize(decoder: &mut EventDecoder<'_>, maximum: usize, field: &str) -> Result<usize> {
    let text = decoder.text(field)?;
    canonical_usize(text, maximum, field)
}

/// 解码按名称/值交替排列的字段数组。
///
/// `array_replace` 为 true 时要求名称严格等于连续数组下标，否则要求名称严格递增。
/// 返回 map 拥有全部字段值；重复、乱序或数量越界使整条事件失败。
fn decode_event_fields(decoder: &mut EventDecoder<'_>, array_replace: bool) -> Result<Fields> {
    let count = decoder
        .array_length()
        .filter(|count| count % 2 == 0 && count / 2 <= MAX_FIELDS as u64)
        .ok_or_else(|| Error::field(Code::Corrupt, "fields"))?;
    let field_count = usize::try_from(count / 2).map_err(|_| Error::field(Code::Corrupt, "fields"))?;
    let mut fields = Fields::new();
    let mut previous = None;
    for index in 0..field_count {
        let name = decoder.text("fields")?;
        let value = decoder.bytes("fields")?;
        let invalid_order = if array_replace {
            array_field_index(name, field_count) != Some(index)
        } else {
            previous.is_some_and(|previous| previous >= name)
        };
        if name.is_empty() || invalid_order {
            return Err(Error::field(Code::Corrupt, "fields"));
        }
        previous = Some(name);
        if fields.insert(name.to_owned(), value).is_some() {
            return Err(Error::field(Code::Corrupt, "fields"));
        }
    }
    Ok(fields)
}

impl<'a> EventDecoder<'a> {
    /// 报告 decoder 是否已精确消费整个 payload。
    fn done(&self) -> bool {
        self.offset == self.payload.len()
    }

    /// 读取 MessagePack 数组头并返回元素数量；类型错误或截断返回 `None`。
    fn array_length(&mut self) -> Option<u64> {
        let code = self.code()?;
        match code {
            0x90..=0x9f => Some(u64::from(code & 0x0f)),
            0xdc => {
                let value = self.take(2)?;
                Some(u64::from(u16::from_be_bytes([value[0], value[1]])))
            }
            0xdd => {
                let value = self.take(4)?;
                Some(u64::from(u32::from_be_bytes([value[0], value[1], value[2], value[3]])))
            }
            _ => None,
        }
    }

    /// 读取 `field` 对应的字符串/二进制值并验证 UTF-8。
    ///
    /// 返回切片直接借用原始 payload，生命周期由 `'a` 约束。
    fn text(&mut self, field: &str) -> Result<&'a str> {
        let value = self.raw_bytes().ok_or_else(|| Error::field(Code::Corrupt, field))?;
        std::str::from_utf8(value).map_err(|_| Error::field(Code::Corrupt, field))
    }

    /// 读取 `field` 对应的字符串/二进制值并复制为拥有型字节。
    fn bytes(&mut self, field: &str) -> Result<Vec<u8>> {
        self.raw_bytes().map(<[u8]>::to_vec).ok_or_else(|| Error::field(Code::Corrupt, field))
    }

    /// 读取一个 MessagePack str/bin 载荷并返回对原始 payload 的零拷贝切片。
    fn raw_bytes(&mut self) -> Option<&'a [u8]> {
        let code = self.code()?;
        let length = match code {
            0xa0..=0xbf => u64::from(code & 0x1f),
            0xc4 | 0xd9 => u64::from(self.take(1)?[0]),
            0xc5 | 0xda => {
                let value = self.take(2)?;
                u64::from(u16::from_be_bytes([value[0], value[1]]))
            }
            0xc6 | 0xdb => {
                let value = self.take(4)?;
                u64::from(u32::from_be_bytes([value[0], value[1], value[2], value[3]]))
            }
            _ => return None,
        };
        self.take(length)
    }

    /// 读取当前位置的一个 MessagePack 类型字节并推进 offset。
    fn code(&mut self) -> Option<u8> {
        let value = *self.payload.get(self.offset)?;
        self.offset += 1;
        Some(value)
    }

    /// 从当前位置借用 `length` 字节并推进 offset；平台转换、加法或边界失败返回 `None`。
    fn take(&mut self, length: u64) -> Option<&'a [u8]> {
        let length = usize::try_from(length).ok()?;
        let end = self.offset.checked_add(length)?;
        let value = self.payload.get(self.offset..end)?;
        self.offset = end;
        Some(value)
    }
}

#[cfg(test)]
#[path = "../../tests/internal/catalog/event.rs"]
mod tests;

File: sdk/go/catalog/event.go
package catalog

import (
	"strings"

	verdandi "github.com/eosforge/verdandi/sdk/go"
)

type eventKind uint8

const (
	eventReplace eventKind = iota + 1
	eventPatch
	eventDelete
)

type catalogEvent struct {
	kind         eventKind
	path         Path
	revision     uint64
	baseRevision uint64
	valueKind    Kind
	encodedBytes int
	fields       verdandi.Fields
}

type eventDecoder struct {
	payload string
	offset  int
}

// decodeEvent 解码并校验单条 Catalog Pub/Sub 通知。
// payload 是 Redis 返回的 MessagePack 原始消息；expectedPath 是订阅者绑定的目标，
// 用于拒绝投递到错误频道或成员的消息；maximumBytes 限制解码后完整值的逻辑大小。
// 返回值只会包含通过协议版本、目标、revision、字段顺序和容量校验的完整事件。
func decodeEvent(payload string, expectedPath Path, maximumBytes int) (catalogEvent, error) {
	// 先按最坏编码开销拒绝异常大消息，避免在解析恶意负载前占用过多 CPU。
	if len(payload) > maximumEventPayload(maximumBytes) {
		return catalogEvent{}, newError(verdandi.CodeCapacity, "notification", 0, nil)
	}
	decoder := eventDecoder{payload: payload}
	count, ok := decoder.arrayLength()
	if !ok || count < 4 {
		return catalogEvent{}, newError(verdandi.CodeCorrupt, "notification", 0, nil)
	}
	protocol, ok := decoder.text()
	if !ok || protocol != "v1" {
		return catalogEvent{}, newError(verdandi.CodeProtocol, "protocol", 0, nil)
	}
	operation, ok := decoder.text()
	if !ok {
		return catalogEvent{}, newError(verdandi.CodeCorrupt, "operation", 0, nil)
	}
	member, ok := decoder.text()
	if !ok || !expectedPath.matchesMember(member) {
		return catalogEvent{}, newError(verdandi.CodeTarget, "path", 0, nil)
	}

	// 公共头部确认后再按操作类型解析固定 ABI，任何分支都必须恰好消费全部元素。
	event := catalogEvent{path: expectedPath}
	var err error
	switch operation {
	case "replace":
		if count != 7 {
			return catalogEvent{}, newError(verdandi.CodeCorrupt, "notification", 0, nil)
		}
		event.kind = eventReplace
		event.revision, err = decodeEventRevision(&decoder, "@revision")
		if err != nil {
			return catalogEvent{}, err
		}
		kindText, kindOK := decoder.text()
		event.valueKind, kindOK = parseKind(kindText)
		if !kindOK {
			return catalogEvent{}, newError(verdandi.CodeCorrupt, "@kind", event.revision, nil)
		}
		bytesText, bytesOK := decoder.text()
		if !bytesOK {
			return catalogEvent{}, newError(
				verdandi.CodeCorrupt, "@encoded_bytes", event.revision, nil,
			)
		}
		event.encodedBytes, err = parseInteger(bytesText, "@encoded_bytes", maximumBytes)
		if err != nil {
			return catalogEvent{}, err
		}
		event.fields, err = decodeEventFields(&decoder, event.valueKind == Array)
		if err != nil {
			return catalogEvent{}, err
		}
		_, actual, validateErr := validateValue(event.valueKind, event.fields, maximumBytes)
		if validateErr != nil || actual != event.encodedBytes {
			return catalogEvent{}, newError(
				verdandi.CodeCorrupt, "@encoded_bytes", event.revision, validateErr,
			)
		}
	case "patch":
		if count != 8 {
			return catalogEvent{}, newError(verdandi.CodeCorrupt, "notification", 0, nil)
		}
		event.kind = eventPatch
		event.baseRevision, err = decodeEventRevision(&decoder, "@base_revision")
		if err != nil {
			return catalogEvent{}, err
		}
		event.revision, err = decodeEventRevision(&decoder, "@revision")
		if err != nil || event.revision <= event.baseRevision {
			return catalogEvent{}, newError(
				verdandi.CodeCorrupt, "@revision", event.revision, err,
			)
		}
		kindText, kindOK := decoder.text()
		event.valueKind, kindOK = parseKind(kindText)
		if !kindOK || event.valueKind == Value {
			return catalogEvent{}, newError(verdandi.CodeCorrupt, "@kind", event.revision, nil)
		}
		bytesText, bytesOK := decoder.text()
		if !bytesOK {
			return catalogEvent{}, newError(
				verdandi.CodeCorrupt, "@encoded_bytes", event.revision, nil,
			)
		}
		event.encodedBytes, err = parseInteger(bytesText, "@encoded_bytes", maximumBytes)
		if err != nil {
			return catalogEvent{}, err
		}
		event.fields, err = decodeEventFields(&decoder, false)
		if err != nil {
			return catalogEvent{}, err
		}
		if _, err := validatePatchFields(event.fields, maximumBytes); err != nil {
			return catalogEvent{}, err
		}
	case "delete":
		if count != 4 {
			return catalogEvent{}, newError(verdandi.CodeCorrupt, "notification", 0, nil)
		}
		event.kind = eventDelete
		event.revision, err = decodeEventRevision(&decoder, "@revision")
		if err != nil {
			return catalogEvent{}, err
		}
	default:
		return catalogEvent{}, newError(verdandi.CodeProtocol, "operation", 0, nil)
	}
	if !decoder.done() {
		return catalogEvent{}, newError(verdandi.CodeCorrupt, "notification", 0, nil)
	}
	return event, nil
}

// maximumEventPayload 计算通知的保守线长上限。
// maximumBytes 是值内容上限；返回值额外容纳字段名称、MessagePack 容器头和协议元数据。
func maximumEventPayload(maximumBytes int) int {
	fields := maximumBytes
	if fields > maximumFields {
		fields = maximumFields
	}
	return maximumBytes + fields*10 + 1024
}

// decodeEventRevision 从 decoder 读取一个正 revision。
// field 仅用于精确标记错误来源；成功时返回协议允许范围内的版本号。
func decodeEventRevision(decoder *eventDecoder, field string) (uint64, error) {
	text, ok := decoder.text()
	if !ok {
		return 0, newError(verdandi.CodeCorrupt, field, 0, nil)
	}
	revision, err := parseRevision(text, false)
	if err != nil {
		return 0, newError(verdandi.CodeCorrupt, field, 0, err)
	}
	return revision, nil
}

// decodeEventFields 解码按名称/值交替排列的字段数组。
// arrayReplace 为 true 时要求名称严格为连续数组下标；否则要求字段名严格递增。
// 返回的字段名和值由本次事件私有存储承载，值切片容量被截断到长度，调用者追加时
// 不会覆盖相邻字段；任何结构、顺序或数量错误都会使整条事件失败。
func decodeEventFields(decoder *eventDecoder, arrayReplace bool) (verdandi.Fields, error) {
	count, ok := decoder.arrayLength()
	if !ok || count%2 != 0 || count/2 > maximumFields {
		return nil, newError(verdandi.CodeCorrupt, "fields", 0, nil)
	}
	fieldCount := int(count / 2)
	remaining := len(decoder.payload) - decoder.offset
	nameCapacity := fieldCount * 16
	if nameCapacity > remaining {
		nameCapacity = remaining
	}
	// 字段名和值分别写入共享连续存储；暴露的值切片把容量截断到自身长度，
	// 从而兼顾低分配和字段间的追加隔离。
	var names strings.Builder
	names.Grow(nameCapacity)
	valueStorage := make([]byte, 0, remaining)
	fields := make(verdandi.Fields, fieldCount)
	previous := ""
	for expectedIndex := range fieldCount {
		name, nameOK := decoder.text()
		value, valueOK := decoder.rawBytes()
		if !nameOK || !valueOK || name == "" {
			return nil, newError(verdandi.CodeCorrupt, "fields", 0, nil)
		}
		if arrayReplace {
			index, canonical := arrayFieldIndex(name, fieldCount)
			if !canonical || index != expectedIndex {
				return nil, newError(verdandi.CodeCorrupt, name, 0, nil)
			}
		} else if previous != "" && previous >= name {
			return nil, newError(verdandi.CodeCorrupt, name, 0, nil)
		}
		previous = name
		nameOffset := names.Len()
		names.WriteString(name)
		nameStorage := names.String()
		valueOffset := len(valueStorage)
		valueStorage = append(valueStorage, value...)
		valueEnd := len(valueStorage)
		fields[nameStorage[nameOffset:]] = valueStorage[valueOffset:valueEnd:valueEnd]
	}
	return fields, nil
}

// done 报告 decoder 是否已经精确消费整个 payload。
func (decoder *eventDecoder) done() bool {
	return decoder.offset == len(decoder.payload)
}

// arrayLength 读取 MessagePack 数组头并返回元素数量。
// 返回 false 表示输入耗尽、长度头不完整或下一个值不是数组。
func (decoder *eventDecoder) arrayLength() (uint64, bool) {
	code, ok := decoder.code()
	if !ok {
		return 0, false
	}
	switch {
	case code >= 0x90 && code <= 0x9f:
		return uint64(code & 0x0f), true
	case code == 0xdc:
		value, ok := decoder.take(2)
		if !ok {
			return 0, false
		}
		return uint64(value[0])<<8 | uint64(value[1]), true
	case code == 0xdd:
		value, ok := decoder.take(4)
		if !ok {
			return 0, false
		}
		return uint64(value[0])<<24 | uint64(value[1])<<16 |
			uint64(value[2])<<8 | uint64(value[3]), true
	default:
		return 0, false
	}
}

// text 读取一个 MessagePack 字符串或二进制值并返回其零拷贝字符串视图。
func (decoder *eventDecoder) text() (string, bool) {
	return decoder.rawBytes()
}

// rawBytes 读取 MessagePack 字符串或二进制载荷。
// 返回的字符串直接引用 payload；调用方不得假定其生命周期长于 decoder.payload。
func (decoder *eventDecoder) rawBytes() (string, bool) {
	code, ok := decoder.code()
	if !ok {
		return "", false
	}
	var length int
	switch {
	case code >= 0xa0 && code <= 0xbf:
		length = int(code & 0x1f)
	case code == 0xc4 || code == 0xd9:
		value, ok := decoder.take(1)
		if !ok {
			return "", false
		}
		length = int(value[0])
	case code == 0xc5 || code == 0xda:
		value, ok := decoder.take(2)
		if !ok {
			return "", false
		}
		length = int(value[0])<<8 | int(value[1])
	case code == 0xc6 || code == 0xdb:
		value, ok := decoder.take(4)
		if !ok {
			return "", false
		}
		encodedLength := uint64(value[0])<<24 | uint64(value[1])<<16 |
			uint64(value[2])<<8 | uint64(value[3])
		if encodedLength > uint64(^uint(0)>>1) {
			return "", false
		}
		length = int(encodedLength)
	default:
		return "", false
	}
	return decoder.take(length)
}

// code 读取当前位置的一个 MessagePack 类型字节并推进 offset。
func (decoder *eventDecoder) code() (byte, bool) {
	if decoder.offset >= len(decoder.payload) {
		return 0, false
	}
	value := decoder.payload[decoder.offset]
	decoder.offset++
	return value, true
}

// take 从当前位置取得 length 字节的零拷贝字符串视图并推进 offset。
// length 为负数或超过剩余负载时返回 false，且不推进位置。
func (decoder *eventDecoder) take(length int) (string, bool) {
	remaining := len(decoder.payload) - decoder.offset
	if length < 0 || length > remaining {
		return "", false
	}
	end := decoder.offset + length
	value := decoder.payload[decoder.offset:end]
	decoder.offset = end
	return value, true
}

// eventString 把 go-redis 可能返回的 string 或 []byte 统一为字符串。
// 其他类型返回 false，避免用格式化转换掩盖协议损坏。
func eventString(value any) (string, bool) {
	switch value := value.(type) {
	case string:
		return value, true
	case []byte:
		return string(value), true
	default:
		return "", false
	}
}

// eventBytes 把 go-redis 可能返回的 string 或 []byte 复制为调用方独占字节切片。
// 复制用于隔离驱动缓冲区；其他类型返回 false。
func eventBytes(value any) ([]byte, bool) {
	switch value := value.(type) {
	case string:
		return []byte(value), true
	case []byte:
		return append([]byte(nil), value...), true
	default:
		return nil, false
	}
}
